medcat/cat.py

```diff
@@ -467,64 +466,64 @@ def _print_stats(self,
 
     def train(self,
               data_iterator: Iterable,
+              nepochs: int = 1,
               fine_tune: bool = True,
               progress_print: int = 1000,
               checkpoint: Optional[Checkpoint] = None,
-              resume_from_checkpoint: bool = False) -> None:
+              is_resumed: bool = False) -> None:
         """ Runs training on the data, note that the maximum length of a line
         or document is 1M characters. Anything longer will be trimmed.
 
         Args:
             data_iterator (Iterable):
                 Simple iterator over sentences/documents, e.g. a open file
                 or an array or anything that we can use in a for loop.
+            nepochs (int):
+                Number of epochs for which to run the training.
             fine_tune (bool):
                 If False old training will be removed.
             progress_print (int):
                 Print progress after N lines.
             checkpoint (Optional[medcat.utils.checkpoint.Checkpoint]):
-                The medcat checkpoint object
-            resume_from_checkpoint (bool):
+                The MedCAT checkpoint object
+            is_resumed (bool):
                 If True resume the previous training; If False, start a fresh new training.
         """
+        if is_resumed:
+            checkpoint = checkpoint or Checkpoint.from_latest_training(self.DEFAULT_TRAIN_CKPT_BASE_DIR)
+            self.log.info(f"Resume training on the most recent checkpoint at {checkpoint.dir_path}...")
+            self.cdb = checkpoint.restore_latest_cdb()
+        else:
+            checkpoint = checkpoint or Checkpoint(dir_path=os.path.join(self.DEFAULT_TRAIN_CKPT_BASE_DIR, str(int(time.time()))))
+            if not fine_tune:
+                self.log.info("Removing old training data!")
+                self.cdb.reset_training()
+            self.log.info(f"Start new training and checkpoints will be saved at {checkpoint.dir_path}...")
+
+        latest_trained_step = checkpoint.count
+        epochal_data_iterator = chain.from_iterable(repeat(data_iterator, nepochs))
+        for line in islice(epochal_data_iterator, checkpoint.count, None):
```

Comment:
latest_trained_step is not counting all lines in a document, and here you are moving forward latest_trained_step including all lines

Reply:
Well spotted. Thought of less ckpts minimising I/O and saving time. `latest_trained_step` is now updated regardless of empty lines.